# Context

* Consider there is a submarine and a war going on between two countries. Submarine of a country is going to another country. The enemy country has placed some mines under water. There can also be rocks in the ocean. Our job is to make a model to identify if there's a rock or a mine underneath the submarine.

* Submarine sends a sound signal which is known as Sonar and detects the object details.

* In real cases, the data has been collected by sending and receiving the sonar signals to and from the metal and a rock. The metal cylinder represents the mine and rock is the ground.

# Methodology

* Collection of Data
* Pre-process the Data
* Train-test split
* Feed it to a ML model
* Train our model
* Test our model

## Importing the dependencies

In [220]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

## Uploading the dataset

In [221]:
# loading the dataset into DataFrame

sonar_data_df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Machine Learning Projects/Sonar - Rock vs Mine Prediction/Sonar data.csv', header = None)

## Exploring the Data

In [222]:
sonar_data_df.head()

,0,1,2,3,4,5,6,7,8,9,...,51,52,53,54,55,56,57,58,59,60
0,0.0200,0.0371,0.0428,0.0207,0.0954,0.0986,0.1539,0.1601,0.3109,0.2111,...,0.0027,0.0065,0.0159,0.0072,0.0167,0.0180,0.0084,0.0090,0.0032,R
1,0.0453,0.0523,0.0843,0.0689,0.1183,0.2583,0.2156,0.3481,0.3337,0.2872,...,0.0084,0.0089,0.0048,0.0094,0.0191,0.0140,0.0049,0.0052,0.0044,R
2,0.0262,0.0582,0.1099,0.1083,0.0974,0.2280,0.2431,0.3771,0.5598,0.6194,...,0.0232,0.0166,0.0095,0.0180,0.0244,0.0316,0.0164,0.0095,0.0078,R
3,0.0100,0.0171,0.0623,0.0205,0.0205,0.0368,0.1098,0.1276,0.0598,0.1264,...,0.0121,0.0036,0.0150,0.0085,0.0073,0.0050,0.0044,0.0040,0.0117,R
4,0.0762,0.0666,0.0481,0.0394,0.0590,0.0649,0.1209,0.2467,0.3564,0.4459,...,0.0031,0.0054,0.0105,0.0110,0.0015,0.0072,0.0048,0.0107,0.0094,R


### Observations
* Final column has R and M which represents Rock and Mine respectively.


In [223]:
sonar_data.shape

(207, 61)

In [224]:
sonar_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 61 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       208 non-null    float64
 1   1       208 non-null    float64
 2   2       208 non-null    float64
 3   3       208 non-null    float64
 4   4       208 non-null    float64
 5   5       208 non-null    float64
 6   6       208 non-null    float64
 7   7       208 non-null    float64
 8   8       208 non-null    float64
 9   9       208 non-null    float64
 10  10      208 non-null    float64
 11  11      208 non-null    float64
 12  12      208 non-null    float64
 13  13      208 non-null    float64
 14  14      208 non-null    float64
 15  15      208 non-null    float64
 16  16      208 non-null    float64
 17  17      208 non-null    float64
 18  18      208 non-null    float64
 19  19      208 non-null    float64
 20  20      208 non-null    float64
 21  21      208 non-null    float64
 22  22

### Statistical Measures

In [225]:
sonar_data_df.describe()

,0,1,2,3,4,5,6,7,8,9,...,50,51,52,53,54,55,56,57,58,59
count,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,...,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000
mean,0.029164,0.038437,0.043832,0.053892,0.075202,0.104570,0.121747,0.134799,0.178003,0.208259,...,0.016069,0.013420,0.010709,0.010941,0.009290,0.008222,0.007820,0.007949,0.007941,0.006507
std,0.022991,0.032960,0.038428,0.046528,0.055552,0.059105,0.061788,0.085152,0.118387,0.134416,...,0.012008,0.009634,0.007060,0.007301,0.007088,0.005736,0.005785,0.006470,0.006181,0.005031
min,0.001500,0.000600,0.001500,0.005800,0.006700,0.010200,0.003300,0.005500,0.007500,0.011300,...,0.000000,0.000800,0.000500,0.001000,0.000600,0.000400,0.000300,0.000300,0.000100,0.000600
25%,0.013350,0.016450,0.018950,0.024375,0.038050,0.067025,0.080900,0.080425,0.097025,0.111275,...,0.008425,0.007275,0.005075,0.005375,0.004150,0.004400,0.003700,0.003600,0.003675,0.003100
50%,0.022800,0.030800,0.034300,0.044050,0.062500,0.092150,0.106950,0.112100,0.152250,0.182400,...,0.013900,0.011400,0.009550,0.009300,0.007500,0.006850,0.005950,0.005800,0.006400,0.005300
75%,0.035550,0.047950,0.057950,0.064500,0.100275,0.134125,0.154000,0.169600,0.233425,0.268700,...,0.020825,0.016725,0.014900,0.014500,0.012100,0.010575,0.010425,0.010350,0.010325,0.008525
max,0.137100,0.233900,0.305900,0.426400,0.401000,0.382300,0.372900,0.459000,0.682800,0.710600,...,0.100400,0.070900,0.039000,0.035200,0.044700,0.039400,0.035500,0.044000,0.036400,0.043900


* ```value_counts()``` is used to count the number of categories of a categorical variable.

In [226]:
sonar_data_df[60].value_counts()

,count
60,
M,111
R,97


* As both the categories have almost equal instances, the accuracy of model is not compromised much.
* More the data, the better is the accuracy.

### Grouping the data for Mine and Rock

In [227]:
sonar_data_df.groupby(60).mean()

,0,1,2,3,4,5,6,7,8,9,...,50,51,52,53,54,55,56,57,58,59
60,,,,,,,,,,,,,,,,,,,,,
M,0.034989,0.045544,0.050720,0.064768,0.086715,0.111864,0.128359,0.149832,0.213492,0.251022,...,0.019352,0.016014,0.011643,0.012185,0.009923,0.008914,0.007825,0.009060,0.008695,0.006930
R,0.022498,0.030303,0.035951,0.041447,0.062028,0.096224,0.114180,0.117596,0.137392,0.159325,...,0.012311,0.010453,0.009640,0.009518,0.008567,0.007430,0.007814,0.006677,0.007078,0.006024


* ```groupby()``` helps us identify if the categories of a categorical variable are significantly different or not.
* We found that for each category there is a significant difference of the mean values for all the features.

## Separating the Data and Labels

In [228]:
X = sonar_data_df.drop(columns = 60, axis = 1)
y = sonar_data_df[60]

In [229]:
X.head()

,0,1,2,3,4,5,6,7,8,9,...,50,51,52,53,54,55,56,57,58,59
0,0.0200,0.0371,0.0428,0.0207,0.0954,0.0986,0.1539,0.1601,0.3109,0.2111,...,0.0232,0.0027,0.0065,0.0159,0.0072,0.0167,0.0180,0.0084,0.0090,0.0032
1,0.0453,0.0523,0.0843,0.0689,0.1183,0.2583,0.2156,0.3481,0.3337,0.2872,...,0.0125,0.0084,0.0089,0.0048,0.0094,0.0191,0.0140,0.0049,0.0052,0.0044
2,0.0262,0.0582,0.1099,0.1083,0.0974,0.2280,0.2431,0.3771,0.5598,0.6194,...,0.0033,0.0232,0.0166,0.0095,0.0180,0.0244,0.0316,0.0164,0.0095,0.0078
3,0.0100,0.0171,0.0623,0.0205,0.0205,0.0368,0.1098,0.1276,0.0598,0.1264,...,0.0241,0.0121,0.0036,0.0150,0.0085,0.0073,0.0050,0.0044,0.0040,0.0117
4,0.0762,0.0666,0.0481,0.0394,0.0590,0.0649,0.1209,0.2467,0.3564,0.4459,...,0.0156,0.0031,0.0054,0.0105,0.0110,0.0015,0.0072,0.0048,0.0107,0.0094


In [230]:
y.head()

,60
0,R
1,R
2,R
3,R
4,R


## Splitting the data into Training and Test data

In [231]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.1, stratify = y, random_state = 1)


* ```test_size = 0.1``` means 10% of the dataset is test data and rest 90% dataset is training data.
* ```stratify``` ensure the equal splitting of R and M for test and train. It prevents skewing.
* ```random_state``` is used for specifying the splitting pattern.

In [232]:
print(X.shape, X_train.shape, X_test.shape)

(208, 60) (187, 60) (21, 60)


## Training the Model

### Logistic Regression Model
* This model works the best in the binary classification problem. The problem is to predict if the detected surface is a rock or a mine.
* This is a Supervised Learning model.

In [233]:
model = LogisticRegression()

In [234]:
model.fit(X_train, y_train)

LogisticRegression()

In [235]:
# predicting the training data
X_train_prediction = model.predict(X_train)

In [236]:
# calculating the accuracy_score
training_data_accuracy = accuracy_score(X_train_prediction, y_train)
print(f'Accuracy on training data: {training_data_accuracy}')

Accuracy on training data: 0.8342245989304813


In [237]:
# predicting the testdata
X_test_prediction = model.predict(X_test)


In [238]:
# calculating the accuracy_score
testing_data_accuracy = accuracy_score(X_test_prediction, y_test)
print(f'Accuracy on testing data: {testing_data_accuracy}')

Accuracy on testing data: 0.7619047619047619


### Random Forest Algorithm
* This model can capture non-linear relationships which the logistic regression cannot capture.

In [239]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report

In [240]:
# n_estimators is the number of trees in the forest
rf_model = RandomForestClassifier(n_estimators=100, random_state=42)

In [241]:
# Fitting the model
rf_model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [242]:
# making predictions
y_pred = rf_model.predict(X_test)

In [243]:
print(f'Accuracy of Random Forest Classifier is: {accuracy_score(y_test, y_pred)}')

Accuracy of Random Forest Classifier is: 0.7619047619047619


In [244]:
print("Classification Report:\n", classification_report(y_test, y_pred))

Classification Report:
               precision    recall  f1-score   support

           M       0.75      0.82      0.78        11
           R       0.78      0.70      0.74        10

    accuracy                           0.76        21
   macro avg       0.76      0.76      0.76        21
weighted avg       0.76      0.76      0.76        21



### Random Forest using Hyperparameter Tuning

In [245]:
# Trying these parameters to prevent overfitting
rf_model_hp = RandomForestClassifier(
    n_estimators=100,
    max_depth=10,        # Limits how deep the trees grow
    min_samples_split=5, # Requires more data to make a "decision" branch
    max_features='sqrt', # Limits how many frequencies each tree looks at
    random_state=42
)


In [246]:
# Fitting the model
rf_model_hp.fit(X_train, y_train)

RandomForestClassifier(max_depth=10, min_samples_split=5, random_state=42)

In [247]:
# making predictions
y_pred_hp = rf_model_hp.predict(X_test)

In [248]:
print(f'Accuracy of Random Forest Classifier is: {accuracy_score(y_test, y_pred_hp)}')

Accuracy of Random Forest Classifier is: 0.7142857142857143


In [249]:
print("Classification Report:\n", classification_report(y_test, y_pred_hp))

Classification Report:
               precision    recall  f1-score   support

           M       0.69      0.82      0.75        11
           R       0.75      0.60      0.67        10

    accuracy                           0.71        21
   macro avg       0.72      0.71      0.71        21
weighted avg       0.72      0.71      0.71        21



## Observations:
* The accuracy is reducing due to the 'small data effect'.
* The sonar data contains just 208 rows of data. This is a small dataset.

## Making a Predictive System

In [250]:
input_data = (0.0283,0.0599,0.0656,0.0229,0.0839,0.1673,0.1154,0.1098,0.1370,0.1767,0.1995,0.2869,0.3275,0.3769,0.4169,0.5036,0.6180,0.8025,0.9333,0.9399,0.9275,0.9450,0.8328,0.7773,0.7007,0.6154,0.5810,0.4454,0.3707,0.2891,0.2185,0.1711,0.3578,0.3947,0.2867,0.2401,0.3619,0.3314,0.3763,0.4767,0.4059,0.3661,0.2320,0.1450,0.1017,0.1111,0.0655,0.0271,0.0244,0.0179,0.0109,0.0147,0.0170,0.0158,0.0046,0.0073,0.0054,0.0033,0.0045,0.0079)

In [251]:
# changing the input_data to a numpy array -> faster calculations
input_data_as_numpy_array = np.asarray(input_data)

In [252]:
# reshape the numpy array as we are predicting for one instance
input_data_reshaped = input_data_as_numpy_array.reshape(1, 60)

prediction = model.predict(input_data_reshaped) # output is a list
print(prediction)

if (prediction[0] == 'R'):
  print('The object is a Rock')
else:
  print('The object is a Mine')

['M']
The object is a Mine
